e2e/e2e_partition_test.go

```diff
@@ -1,14 +1,133 @@
+// Correctness Analysis of Istanbul Byzantine Fault Tolerance (https://arxiv.org/pdf/1901.07160.pdf).
+// Specific problem this test is emulating is described in Chapter 7.1, Case 1.
+// Summary of the problem is that there are not enough nodes to lock on single proposal,
+// due to some issues where nodes being unable to deliver messages to all of the peers.
+// Therefore nodes are split into two subsets locked on different proposals.
+// Simulating one node from larger subset as faulty one, results in being unable to reach a consensus on a single proposal and that's what liveness issue is about.
+// This test creates a single cluster of 5 nodes, but instead of letting all the peers communicate with each other,
+// it routes messages only to specific nodes and induces that nodes lock on different proposal.
+// In the round=1, it marks one node as faulty (meaning that it doesn't takes part in gossiping).
+func TestE2E_Partition_Liveness(t *testing.T) {
+	const nodesCnt = 5
+	round0 := roundMetadata{
+		round: 0,
+		// induce locking A_3 and A_4 on one proposal
+		routingMap: map[sender]receivers{
+			"A_0": {"A_3", "A_4"},
+			"A_3": {"A_0", "A_3", "A_4"},
+			"A_4": {"A_3", "A_4"},
+		},
+	}
+	round1 := roundMetadata{
+		round: 1,
+		// induce locking lock A_0 and A_2 on another proposal
+		routingMap: map[sender]receivers{
+			"A_0": {"A_0", "A_2", "A_3", "A_4"},
+			"A_1": {"A_0", "A_2", "A_3", "A_4"},
+			"A_2": {"A_0", "A_1", "A_2", "A_3", "A_4"},
+
+			"A_3": {"A_0", "A_1", "A_2", "A_3", "A_4"},
+			"A_4": {"A_0", "A_1", "A_2", "A_3", "A_4"},
+		},
+	}
+	flowMap := map[uint64]roundMetadata{0: round0, 1: round1}
+
+	faultyNodeId := pbft.NodeID("A_1")
+
+	// If livenessGossipHandler returns false, message should not be transported.
+	livenessGossipHandler := func(senderId, receiverId pbft.NodeID, msg *pbft.MessageReq) bool {
+		if msg.View.Round > 1 || msg.View.Sequence > 2 {
+			// Faulty node is unresponsive after round 1, and all the other nodes are gossiping all the messages.
+			return senderId != faultyNodeId && receiverId != faultyNodeId
+		} else {
+			if msg.View.Round <= 1 && msg.Type == pbft.MessageReq_Commit {
+				// Cut all the commit messages gossiping for round 0 and 1
+				return false
+			}
+			if msg.View.Round == 1 && senderId == faultyNodeId &&
+				(msg.Type == pbft.MessageReq_RoundChange || msg.Type == pbft.MessageReq_Commit) {
+				// Case where we are in round 1 and 2 different nodes will lock the proposal
+				// (consequence of faulty node doesn't gossip round change and commit messages).
+				return false
+			}
+		}
+
+		return shouldGossipBasedOnMsgFlowMap(flowMap, msg, senderId, receiverId)
+	}
+
+	hook := newGenericGossipTransport(livenessGossipHandler)
+
+	c := newPBFTCluster(t, "liveness_issue", "A", nodesCnt, hook)
+	c.Start()
+	defer c.Stop()
+
+	err := c.WaitForHeight(3, 5*time.Minute)
+
+	// Query nodes and make appropriate assertions
+	nodeSubsetA := []pbft.NodeID{"A_0", "A_1", "A_2"}
```

Comment:
Please revert this change as well. We do not need to do an extra check to see if there are two locked proposals, it is enough for me if we log it. This is the idea behind the integration tests, to test some specific multi node situation and wait for the chain to finish. I do not want to over complicate it this way.